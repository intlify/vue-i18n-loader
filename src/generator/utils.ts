import { isObject, isString } from '@intlify/shared'
import { VisitorKeys } from 'eslint-visitor-keys'
import { parseForESLint, VisitorKeys as VISITOR_KEYS } from 'yaml-eslint-parser'

import type { YAMLProgram, YAMLNode } from 'yaml-eslint-parser/lib/ast'

export interface Visitor<N> {
  visitorKeys?: VisitorKeys
  enterNode(node: N, parent: N | null): void
  leaveNode(node: N, parent: N | null): void
}

export function parseYAML(code: string, options = {}): YAMLProgram {
  const parserOptions = Object.assign(
    {
      filePath: '<input>',
      ecmaVersion: 2019
    },
    options,
    {
      loc: true,
      range: true,
      raw: true,
      tokens: true,
      comment: true,
      eslintVisitorKeys: true,
      eslintScopeManager: true
    }
  )

  return parseForESLint(code, parserOptions).ast
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fallbackKeysFilter(this: any, key: string) {
  let value = null
  return (
    key !== 'comments' &&
    key !== 'sourceType' &&
    key !== 'loc' &&
    key !== 'parent' &&
    key !== 'range' &&
    key !== 'tokens' &&
    (value = this[key]) !== null &&
    isObject(value) &&
    (isString(value.type) || Array.isArray(value))
  )
}

function getFallbackKeys(node: YAMLNode): string[] {
  return Object.keys(node).filter(fallbackKeysFilter, node)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isNode(x: any): x is YAMLNode {
  return x !== null && isObject(x) && isString(x.type)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function* getNodes(node: any, key: string): IterableIterator<YAMLNode> {
  const child = node[key]
  if (Array.isArray(child)) {
    for (const c of child) {
      if (isNode(c)) {
        yield c
      }
    }
  } else if (isNode(child)) {
    yield child
  }
}

function getKeys(node: YAMLNode, visitorKeys?: VisitorKeys) {
  const keys = (visitorKeys || VISITOR_KEYS)[node.type] || getFallbackKeys(node)
  return keys.filter((key: string) => !getNodes(node, key).next().done)
}

function traverse(
  node: YAMLNode,
  parent: YAMLNode | null,
  visitor: Visitor<YAMLNode>
) {
  visitor.enterNode(node, parent)
  const keys = getKeys(node, visitor.visitorKeys)
  for (const key of keys) {
    for (const child of getNodes(node, key)) {
      traverse(child, node, visitor)
    }
  }
  visitor.leaveNode(node, parent)
}

export function traverseNodes(
  node: YAMLNode,
  visitor: Visitor<YAMLNode>
): void {
  traverse(node, null, visitor)
}
