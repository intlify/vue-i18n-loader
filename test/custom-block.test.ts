import { bundleAndRun } from './utils'
import { createMessageContext } from '@intlify/runtime'

test('basic', async () => {
  const { module } = await bundleAndRun('basic.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual('hello world!')
})

test('special characters', async () => {
  const { module } = await bundleAndRun('special-char.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual(
    'hello\ngreat\t"world"'
  )
})

test('multiple', async () => {
  const { module } = await bundleAndRun('multiple.vue')
  expect(module.__i18n).toMatchSnapshot()
  expect(module.__i18n.length).toEqual(2)
  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.ja.hello(createMessageContext())).toEqual(
    'こんにちは、世界！'
  )
  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual('hello world!')
})

test('import', async () => {
  const { module } = await bundleAndRun('import.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual('hello world!')
})

test('locale attr', async () => {
  const { module } = await bundleAndRun('locale.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  expect(i18n.resource.hello(createMessageContext())).toEqual(
    'こんにちは、世界！'
  )
})

test('locale attr and basic', async () => {
  const { module } = await bundleAndRun('locale-mix.vue')
  expect(module.__i18n).toMatchSnapshot()
  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  expect(i18n.resource.hello(createMessageContext())).toEqual(
    'こんにちは、世界！'
  )
  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual('hello world!')
})

test('locale attr and import', async () => {
  const { module } = await bundleAndRun('locale-import.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('en')
  expect(i18n.resource.hello(createMessageContext())).toEqual('hello world!')
})

test('yaml', async () => {
  const { module } = await bundleAndRun('yaml.vue')
  expect(module.__i18n).toMatchSnapshot()
  let i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('ja')
  expect(i18n.resource.hello(createMessageContext())).toEqual(
    'こんにちは、世界！'
  )
  i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('en')
  expect(i18n.resource.hello(createMessageContext())).toEqual('hello world!')
})

test('json5', async () => {
  const { module } = await bundleAndRun('json5.vue')
  expect(module.__i18n).toMatchSnapshot()
  const i18n = module.__i18n.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual('hello world!')
})

test('global', async () => {
  const { module } = await bundleAndRun('global.vue')
  expect(module.__i18n).toBeUndefined()
  expect(module.__i18nGlobal).toMatchSnapshot()
  const i18n = module.__i18nGlobal.pop()
  expect(i18n.locale).toEqual('')
  expect(i18n.resource.en.hello(createMessageContext())).toEqual(
    'hello global!'
  )
})

test('global and local', async () => {
  const { module } = await bundleAndRun('global-mix.vue')
  expect(module.__i18n).toMatchSnapshot()
  expect(module.__i18nGlobal).toMatchSnapshot()
  const l = module.__i18n.pop()
  expect(l.locale).toEqual('ja')
  expect(l.resource.hello(createMessageContext())).toEqual('hello local!')
  const g = module.__i18nGlobal.pop()
  expect(g.locale).toEqual('')
  expect(g.resource.en.hello(createMessageContext())).toEqual('hello global!')
})
