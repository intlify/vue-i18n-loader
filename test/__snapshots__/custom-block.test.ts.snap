// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`global 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`global and local 1`] = `
Array [
  Object {
    "locale": "ja",
    "resource": Object {
      "hello": [Function],
    },
  },
]
`;

exports[`global and local 2`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`import 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`json5 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`locale attr 1`] = `
Array [
  Object {
    "locale": "ja",
    "resource": Object {
      "hello": [Function],
    },
  },
]
`;

exports[`locale attr and basic 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
  Object {
    "locale": "ja",
    "resource": Object {
      "hello": [Function],
    },
  },
]
`;

exports[`locale attr and import 1`] = `
Array [
  Object {
    "locale": "en",
    "resource": Object {
      "hello": [Function],
    },
  },
]
`;

exports[`multiple 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
  Object {
    "locale": "",
    "resource": Object {
      "ja": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`special characters 1`] = `
Array [
  Object {
    "locale": "",
    "resource": Object {
      "en": Object {
        "hello": [Function],
      },
    },
  },
]
`;

exports[`yaml 1`] = `
Array [
  Object {
    "locale": "en",
    "resource": Object {
      "hello": [Function],
    },
  },
  Object {
    "locale": "ja",
    "resource": Object {
      "hello": [Function],
    },
  },
]
`;
